PF-updates

pfellerich opened the pull request:
# A few additions and bug fixes

1. pkey.new{} accepts 'dhparam' as named parameter, to provide pre-computed DH parameters to speed up the key generation. The use of 'dhparam' overrides the 'bits' parameter.

2. pkey.new{} uses named parameter 'generator' (which defaults to 2) to specify the DH generator, and not the 'exp' parameter, which defaults to 65537. 

3. pkey.new() accepts an optional passphrase as 4th parameter to allow to load password protected keys.

4. pkcs12.parse(string pkcs12bag[, string passphrase]) added to read a standard PKCS12 file provided as a string. The function returns 3 values (pkey, cert, CA) in their respective object representations.

5. crl:getNextUpdate() always returned 1 as unixtime, bug fixed.

daurnimator commented on src/openssl.c:
```diff
@@ -3103,6 +3105,21 @@ static int pk_new(lua_State *L) {
+				luaL_argerror(L, 1, lua_pushfstring(L, "dhparam must be a PEM encoded string"));
+		}
+
+		/* genarator is the one magic number that can be passed in for DH_generate_params */
```
typo: genarator

daurnimator commented on src/openssl.c:
```diff
@@ -3103,6 +3105,21 @@ static int pk_new(lua_State *L) {
+		 *   `openssl dhparam -2 -out dh-2048.pem -outform PEM 2048`
+		 */
+		if (loadfield(L, 1, "dhparam", LUA_TSTRING, &dhparam)) {
+			if (!dhparam || strlen(dhparam)<200)
```
I feel like we should avoid `strlen` and use length as given by lua.

daurnimator commented on src/openssl.c:
```diff
@@ -3139,9 +3156,33 @@ static int pk_new(lua_State *L) {
+			 * of an error, the bit flags DH_CHECK_P_NOT_SAFE_PRIME or 
+			 * DH_NOT_SUITABLE_GENERATOR are set in *codes.
+			 */
+			if (0==DH_check(dh, &codes) || 0!=codes) {
```
style nit: whitespace around equality operators `0 != codes`

daurnimator commented on src/openssl.c:
```diff
@@ -3047,12 +3047,26 @@ static BIO *getbio(lua_State *L) {
 	return *bio;
 } /* getbio() */
 
+/* not used
```
So this function isn't required? Why is it here and not used?

daurnimator replied:
Ah, I see this was then removed in a subsequent commit. Please rebase.

daurnimator commented on src/openssl.c:
```diff
@@ -7474,6 +7462,65 @@ static int p12_interpose(lua_State *L) {
 
+static int p12_parse(lua_State *L) {
+	/* parse a p12 binary string and return the parts */
+	lua_settop(L,2);
```
style nit: space after `,`

daurnimator replied:
There is no need to use `settop` in this function.

daurnimator commented on src/openssl.c:
```diff
@@ -7474,6 +7462,65 @@ static int p12_interpose(lua_State *L) {
+		return auxL_error(L, auxL_EOPENSSL, "pkcs12.parse");
+
+	/* passphrase is the second parameter */
+	passphrase = luaL_optstring(L, 2, NULL);
```
This (and the `prepsimple`s below) can longjump out, leaking `p12`. 
Do all the argument marshalling and userdata allocations upfront before any openssl allocations.

daurnimator commented on src/openssl.c:
```diff
@@ -7173,6 +7173,14 @@ static int xs_add(lua_State *L) {
+			X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_CRL_CHECK);
+			X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_CRL_CHECK_ALL);
+			X509_STORE_set1_param(store, param);
+			X509_VERIFY_PARAM_free(param);//*/
```
empty comment?

daurnimator commented on src/openssl.c:
```diff
@@ -7173,6 +7173,14 @@ static int xs_add(lua_State *L) {
 				X509_CRL_free(crl_dup);
 				return auxL_error(L, auxL_EOPENSSL, "x509.store:add");
 			}
```
this commit message has a typo. (x5009 vs x509)

daurnimator commented on src/openssl.c:
```diff
@@ -7173,6 +7173,14 @@ static int xs_add(lua_State *L) {
+			X509_VERIFY_PARAM *param = X509_VERIFY_PARAM_new();
+			X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_CRL_CHECK);
+			X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_CRL_CHECK_ALL);
+			X509_STORE_set1_param(store, param);
```
I'm hesistant to do this by default when openssl itself doesn't do it by default.
Instead, could you add bindings to `X509_VERIFY_PARAM_set_flags` so that lua users can do this themselves?

pfellerich commented:
Hi

and thanks for the review.

I fixed the typos and other trivial things, for the remaining ones I do 
have some comments, see below.

Am 23.03.2017 um 08:27 schrieb daurnimator:
>
> ------------------------------------------------------------------------
>
> In src/openssl.c 
> <https://github.com/wahern/luaossl/pull/90#discussion_r107600662>:
>
> > @@ -3103,6 +3105,21 @@ static int pk_new(lua_State *L) {
>   				luaL_argerror(L, 1, lua_pushfstring(L, "%s: invalid curve", id));
>   		}
>   
> +		/* dhparam field can contain a PEM encoded string.
> +		 * That string can be created and written to a file like so:
> +		 *   `openssl dhparam -2 -out dh-2048.pem -outform PEM 2048`
> +		 */
> +		if (loadfield(L, 1, "dhparam", LUA_TSTRING, &dhparam)) {
> +			if (!dhparam || strlen(dhparam)<200)
>
> I feel like we should avoid |strlen| and use length as given by lua.
>
I fully agree, I prefer this as well. However the existing "loadfield" 
function returns a C-string, and I didn't want to make larger changes 
than needed. On the other hand, since lua_tolstring() accepts a NULL 
pointer as length target and guarantees a trailing null byte, the change 
to loadfield() may be trivial enough to do.

> ------------------------------------------------------------------------
>
> In src/openssl.c 
> <https://github.com/wahern/luaossl/pull/90#discussion_r107601495>:
>
> > @@ -7474,6 +7462,65 @@ static int p12_interpose(lua_State *L) {
>   } /* p12_interpose() */
>   
>   
> +static int p12_parse(lua_State *L) {
> +	/* parse a p12 binary string and return the parts */
> +	lua_settop(L,2);
>
> There is no need to use |settop| in this function.
>
Probably a leftover to cope with the optional 2nd parameter and 
collisions when pushing more data on the stack. Removed.
> ------------------------------------------------------------------------
>
> In src/openssl.c 
> <https://github.com/wahern/luaossl/pull/90#discussion_r107601734>:
>
> > +	PKCS12 *p12;
> +
> +	const char *blob, *passphrase;
> +	size_t len;
> +
> +	blob = luaL_checklstring(L, 1, &len);
> +	BIO *bio = BIO_new_mem_buf((void*)blob, len);
> +	if (!bio)
> +		return auxL_error(L, auxL_EOPENSSL, "pkcs12.parse");
> +	p12 = d2i_PKCS12_bio(bio, NULL);
> +	BIO_free(bio);
> +	if (!p12)
> +		return auxL_error(L, auxL_EOPENSSL, "pkcs12.parse");
> +
> +	/* passphrase is the second parameter */
> +	passphrase = luaL_optstring(L, 2, NULL);
>
> This (and the |prepsimple|s below) can longjump out, leaking |p12|.
> Do all the argument marshalling and userdata allocations upfront 
> before any openssl allocations.
>
I refactored the code as you suggested. That is: argument preparations, 
UD allocation, then SSL stuff, and finally return value updates. Updates 
means: replacing userdata by nil. The nice thing is that it works 
because the original code does not only zero out the new userdata, but 
also because the __gc hooks (well, most of them) call the free function 
only if the pointer is set -- that's what I call defensive coding.

> ------------------------------------------------------------------------
>
> In src/openssl.c 
> <https://github.com/wahern/luaossl/pull/90#discussion_r107601795>:
>
> > @@ -7173,6 +7173,14 @@ static int xs_add(lua_State *L) {
>   				X509_CRL_free(crl_dup);
>   				return auxL_error(L, auxL_EOPENSSL, "x509.store:add");
>   			}
> +
> +			/* When a CRL was added, enable CRL checking (entire chain) */
> +			X509_VERIFY_PARAM *param = X509_VERIFY_PARAM_new();
> +			X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_CRL_CHECK);
> +			X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_CRL_CHECK_ALL);
> +			X509_STORE_set1_param(store, param);
> +			X509_VERIFY_PARAM_free(param);//*/
>
> empty comment?
>
No, just my quick code block switcher during testing. Adding a space 
inbetween the  */  comments out the whole block. Can also be used to 
alternate between two code blocks:

    // /*
    block1;
    /*/
    alternative1;
    //*/

I'll remove it.

> ------------------------------------------------------------------------
>
> In src/openssl.c 
> <https://github.com/wahern/luaossl/pull/90#discussion_r107601995>:
>
> > @@ -7173,6 +7173,14 @@ static int xs_add(lua_State *L) {
>   				X509_CRL_free(crl_dup);
>   				return auxL_error(L, auxL_EOPENSSL, "x509.store:add");
>   			}
> +
> +			/* When a CRL was added, enable CRL checking (entire chain) */
> +			X509_VERIFY_PARAM *param = X509_VERIFY_PARAM_new();
> +			X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_CRL_CHECK);
> +			X509_VERIFY_PARAM_set_flags(param, X509_V_FLAG_CRL_CHECK_ALL);
> +			X509_STORE_set1_param(store, param);
>
> I'm hesistant to do this by default when openssl itself doesn't do it 
> by default.
> Instead, could you add bindings to |X509_VERIFY_PARAM_set_flags| so 
> that lua users can do this themselves?
>

I am undecided on this one. I can see your point, but on the other hand, 
looking at the various StackOverflow threads, it seems that this is an 
easily forgotten item. From the Lua side, adding a certificate to the 
context implies using it. Adding a CRL to the context also expresses the 
desire to have the CRL checks activated, hence the automatic setting.
Arguments against this default action are that you might want to add 
several CRLs to the context, in which case the flags get set repeatedly; 
and most prominently, you might want to have limited CRL checks, where 
only the leaf certificates are checked against the CRL (= only 
X509_V_FLAG_CRL_CHECK set).


Another missing item is an additional CRL method to verify a CRL against 
a public key. This will be in the updated code as well.


regards
Pascal

daurnimator commented on src/openssl.c:
```diff
@@ -6866,6 +6866,20 @@ static int xx_sign(lua_State *L) {
+} /* xx_verify() */
+
+
+
```
One too many empty lines.

daurnimator commented:
Please rebase so that you have a clear series of commits.

pfellerich commented:
Just did a git pull && git rebase master && git push origin. Is that
what you wanted?


Am 27.03.2017 um 04:58 schrieb daurnimator:
>
> Please rebase so that you have a clear series of commits.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wahern/luaossl/pull/90#issuecomment-289342581>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANB_R6dcBqFkcd0Swov9G8oTDyGbpBFFks5rpyXvgaJpZM4MjiI0>.
>

daurnimator commented:
> Just did a git pull && git rebase master && git push origin. Is that what you wanted?

I meant more like do a `git rebase -i` and edit your commits so that e.g. they don't add bits that are later removed, that e.g. `pk_pem_password_cb` is never seen, and that the whitespace fixes are added into the original commit that introduced the code.

pfellerich commented:
Rebased 6 commits into 3.

daurnimator commented:
Thanks! However the 2nd commit still contains fixups for the first (in particular `p12_parse`, but also the whitespace fixes).

I'd also like to request you keep the `X509_VERIFY_PARAM_set_flags` part separate (possibly in another PR), as I'm unsure about merging that.

daurnimator commented:
:( You just rebased everything into a single commit.

Also something else I realised: you are using `DH_check` unconditionally. This prevents a user from using an (intentionally) unsafe set of parameters (which I do to test that clients reject unsafe servers!).
The check will need to sit behind a flag.

I'll fix things up now.

daurnimator commented:
I've merged only the parts of this PR I was confident about. The remainder I've left on a branch over here: https://github.com/wahern/luaossl/compare/master...daurnimator:pfellerich-pfupdates-rebased?expand=1
You may want to submit that as a new PR?

daurnimator commented:
@pfellerich I've resolved the merge conflicts with my branch containing your code. Did you still need any of it?
